### PR TITLE
Providing core and node level metrics for request handlers

### PR DESCRIPTION
**Pull request by justinrsweeney**
This reverts the previous attempt to get node level metrics (via coreLevelMetricsEnabled flag) and attempts a different method by allowing for multiple registries for certain metrics and capturing both node and core level metrics.

Note: this has undergone basic testing, but needs more testing before merge. Getting a PR open for initial review.

### CODE REVIEW

**Comment by magibney** [comment]
I haven't tested this yet, but looking over the code this seems like a solid approach. I really only have one minor high-level question/thought:

node-level and core-level seem very clear, and clearly useful. It's nice to support arbitrarily many secondary metrics, but unless we envision actually using support for an arbitrary number of metric registries, it might be cleaner (at least initially) to take a more limited approach, simply accepting one additional "delegate" metric, rather than looping over a set when we expect to only ever have one secondary metric. The only other possible option I could imagine that might justify keeping the "arbitrarily-many-metrics" approach would be implicit configuration to collect metrics at the level of collection (aggregate for cores belonging to the same collection that happen to be on the same node).

To be able to configure more than core-level.+ node-level, we'd probably need to extend explicit configuration options beyond the current implicit "we only care about RequestHandlers" approach. Scaling back in this way should simplify the code further (e.g., allowing to remove all the new imports on `SolrMetricsContext.java`).

I think for a local mod it's fine to stick with implicit configuration here, if we're only interested in RequestHandlers at the node-level, and are interested (or at least not _dis_-interested) in _all_ RequestHandlers.

**Comment by justinrsweeney** [comment]
I went ahead with a bit of a combination. I kept the classes that extend Meter, Counter, Timer to iterate an array of those objects since I think keeping those fairly generic is useful.

I created a new SolrDelegateRegistryMetricsContext class to extend the SolrMetricsContext to override methods to a single additional delegate registry (node level in this case).

Finally, I did make use of this be configurable from solrconfig on a per RequestHandler basis. I think we only really need this for a handful of RequestHandlers so I like the idea of having it be explicitly configurable since we can limit it and also have an easy way to enable/disable if any issue were to arise.

**Comment by justinrsweeney** [comment]
Also, I'm going to create a new PR rolling back the previous node level metric implementation to make this PR easier to read.

**Comment by justinrsweeney** [comment]
This current setup would allow for enabling on a specific request handler via config like:
```
<requestHandler name="/select" class="solr.SearchHandler">
        <!-- default values for query parameters can be specified, these
             will be overridden by parameters in the request
          -->
        <lst name="defaults">
            <str name="echoParams">explicit</str>
            <int name="rows">10</int>
        </lst>
        <!-- add optics to the end of components to run -->
        <arr name="last-components">
            <str>optics</str>
        </arr>

        <bool name="aggregateNodeLevelMetricsEnabled">true</bool>
    </requestHandler>
```

**Comment by noblepaul** [comment]
lemme try to understand the requirements.

 The metrics API should expose 2 different values for each handlers

1. per-core metrics of each handler
2. per node metrics of each handler. This is the cumulative data of all the counters from all the cores.


I am wondering  if we can just fetch all the values and compute the cumulative values just in time and avoid the `aggregateNodeLevelMetricsEnabled` flag altogether. If there is zero extra cost during queries and there is a small cost of adding the values while fetching the metrics ?

**Comment by justinrsweeney** [comment]
> lemme try to understand the requirements.
> 
> The metrics API should expose 2 different values for each handlers
> 
> 1. per-core metrics of each handler
> 2. per node metrics of each handler. This is the cumulative data of all the counters from all the cores.
> 
> I am wondering if we can just fetch all the values and compute the cumulative values just in time and avoid the `aggregateNodeLevelMetricsEnabled` flag altogether. If there is zero extra cost during queries and there is a small cost of adding the values while fetching the metrics ?

Yes, although, for 2 we are looking for both counters and meter (histogram) metrics. The meters are the ones that are the ones that we can't just aggregate after the fact and are the real driver for these changes. The way meter metrics work in Dropwizard doesn't really allow for aggregation after the fact.

I think with these changes the cost of accumulating a second metric is going to be very minimal in the overall processing time and we can configure this on a per request handler basis.

**Comment by magibney** [comment]
There's a little more detail on the original PR wrt the problems of aggregating histogram metrics: https://github.com/fullstorydev/lucene-solr/pull/155#issuecomment-1162966114

**Comment by magibney** [comment]
I think that even if going with an "arbitrarily-many-delegate-metrics" approach, it'd be a good idea to have each metric returned from the `SolrDelegateRegistryMetricsContext` act in and of itself as a "primary metric", as opposed to the current code in which the returned metric is just a shell that delegates to equivalent "backing" metrics. There's a couple of reasons to prefer this approach:
1. we don't know how calling code is going to use these, whether they may call "read" methods on the returned objects directly. iiuc in the current state of the code, these "read" methods would return values suggesting that the metrics had collected nothing at all (b/c all the actual work was delegated to backing metrics).
2. on principle: there really is a distinction between the "primary" metric (updated by calling `super.[writeMethod]()`) and the aggregate metric; maintaining this distinction (even if sticking with looping over arbitrarily many other delegate metrics as opposed to having a single delegate) is probably worthwhile.

Taking this approach makes it easier to treat the returned metrics as "drop-in" replacements, in every respect, for the "plain old" metrics, with no aggregation.

Shifting gears ... we have 5 kinds of metrics in Solr:
1. gauge - very specialized, initializer responsible for writing custom metrics, virtually impossible to aggregate in a generic way
2. counter - straightforward; could be aggregated just-in-time, post-hoc, on-demand, whatever (as @noblepaul suggests; i.e., does not necessarily require modifications to how metrics are _collected_).
3. meter
4. histogram
5. timer (incorporates meter and histogram)

The last three collect in a way that IIUC can't be aggregated post-hoc. (tbh I haven't dived into this, but it seems everyone's in agreement here?). I guess the question then is: given that delegation seems like the only way to go for the last three, is it worth taking a different (post-hoc) approach for `counter`, even if in principle it would be possible to do so?

**Comment by justinrsweeney** [comment]
> I think that even if going with an "arbitrarily-many-delegate-metrics" approach, it'd be a good idea to have each metric returned from the SolrDelegateRegistryMetricsContext act in and of itself as a "primary metric", as opposed to the current code in which the returned metric is just a shell that delegates to equivalent "backing" metrics. There's a couple of reasons to prefer this approach:
> 
> we don't know how calling code is going to use these, whether they may call "read" methods on the returned objects directly. iiuc in the current state of the code, these "read" methods would return values suggesting that the metrics had collected nothing at all (b/c all the actual work was delegated to backing metrics).
> on principle: there really is a distinction between the "primary" metric (updated by calling super.[writeMethod]()) and the aggregate metric; maintaining this distinction (even if sticking with looping over arbitrarily many other delegate metrics as opposed to having a single delegate) is probably worthwhile.
> Taking this approach makes it easier to treat the returned metrics as "drop-in" replacements, in every respect, for the "plain old" metrics, with no aggregation.

This makes sense, going this route it'll actually be easier to do single delegate class extending Meter, Timer, etc instead of arbitrarily many. That'll follow a consistent pattern of primary + one delegate which I think is sufficient. This could change if there was a desire for additional levels of metrics, but we could cross that bridge if that day comes. I'll update based on this.

**Comment by justinrsweeney** [comment]
> Shifting gears ... we have 5 kinds of metrics in Solr:
> 
> gauge - very specialized, initializer responsible for writing custom metrics, virtually impossible to aggregate in a generic way
> counter - straightforward; could be aggregated just-in-time, post-hoc, on-demand, whatever (as @noblepaul suggests; i.e., does not necessarily require modifications to how metrics are collected).
> meter
> histogram
> timer (incorporates meter and histogram)
> The last three collect in a way that IIUC can't be aggregated post-hoc. (tbh I haven't dived into this, but it seems everyone's in agreement here?). I guess the question then is: given that delegation seems like the only way to go for the last three, is it worth taking a different (post-hoc) approach for counter, even if in principle it would be possible to do so?

My understanding is that metrics like histograms, timers, meters do sampling to create a statistically representative histogram. I think attempting to aggregate such values will result in incorrect representations of the data. Solr's own code also makes that point: https://github.com/fullstorydev/lucene-solr/blob/5c4168d3b2e1c05a09b608fac048be26ae8050d4/solr/core/src/java/org/apache/solr/metrics/reporters/solr/SolrReporter.java#L190

My argument for doing Counter in addition to the others is that querying the /admin/metrics API for core level counters when there are a large number of cores results in quite a large JSON response that then needs to be parsed and aggregated. It is possible, but likely more of a strain than just incrementing an additional metric in the first place.

**Comment by magibney** [comment]
> My argument for doing Counter in addition to the others ...

I'm on board with this argument, fwiw. But it's possible that Noble was suggesting doing the post-hoc counter aggregation Solr-side (as opposed to parsing and combining externally, which would indeed be annoying). I imagine this would be possible, but I took a quick look and didn't immediately see an obvious, non-invasive way to accomplish it.

**Comment by noblepaul** [comment]
> My argument for doing Counter in addition to the others is that querying the /admin/metrics API for core level counters when there are a large number of cores results in quite a large JSON response

Agree. they are largely useless in the current form and an aggregate is the best.

>  But it's possible that Noble was suggesting doing the post-hoc counter aggregation Solr-side

Yes.


>    Histograms are difficult / impossible to aggregate, so it may not be 
If they are difficult, we could just add one one extra object per `CoreContainer` and expose them as a separate metrics ? That change would be minimal too. Also , if the data collection is cheap, why don't we collect them anyway instead of an extra flag ?

**Comment by magibney** [comment]
>If they are difficult, we could just add one one extra object per CoreContainer and expose them as a separate metrics ?

If I correctly understand what you're suggesting, I think that's basically what this PR does?

>if the data collection is cheap, why don't we collect them anyway instead of an extra flag ?

I guess that's an open question. I assume it's cheap, but it's not _nothing_. Paradoxically, the best way to A/B test to determine the performance impact of aggregate collection would be to initially introduce the extra flag, even if we remove it later once we've satisfied ourselves that the performance impact is negligible :-)

Another possible reason to treat the node-level aggregate metrics as "opt-in" is that if we're thinking of this as potentially useful upstream, many users of solr have only 1 or 2 collections, so for them, these aggregate metrics would be redundant.

**Comment by justinrsweeney** [comment]
There doesn't seem to be a reasonable way to the the new Meter, Timer, Counter, Histogram classes to be able to do super.foo() for primary and then also do delegate metrics, due to the way these are instantiated for different registries. I tried to go a bit down that path, but at this point I think this is the best we can do where the any getters just return the primary value in case anyone is using those, but delegate metrics still get updated.

As far as the configuration, I do think it is useful to be able to enable/disable in this way without have to change the Solr version and as @magibney mentioned, there is no reason to enable for single core per node cases.

**Comment by magibney** [comment]
> There doesn't seem to be a reasonable way

Yes, I see what you mean. The cleanest alternative pursuing the `super.foo()` approach would be to add variants of `SolrMetricManager.meter(...)` etc. that would support explicitly specifying a`MetricSupplier`. But the way you've done is probably better, at least for now (and has the added benefit of dodging the awkward questions of how supplying explicit `MetricSupplier` would interact with explicitly specified MetricSupplier impls in `solr.xml` metrics config).

File: solr/core/src/java/org/apache/solr/metrics/DelegateRegistryTimer.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  @Override
+  public void update(Duration duration) {
+    primaryTimer.update(duration);
+    delegateTimer.update(duration);
+  }
+
+  @Override
+  public <T> T time(Callable<T> event) throws Exception {
+    return primaryTimer.time(event);
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/metrics/DelegateRegistryTimer.java]: I think you actually want to have `this` handle calls to `time(Callabale)`, `timeSupplier(Supplier)`, `time(Runnable)`, and `Context time()` -- i.e., these four methods should _not_ be overridden.
This is because they turn around and call `update(...)` on `this`, so if you delegate the `time()` calls, the update call goes to the primaryTimer, and `this` never has the opportunity to distribute the update to primary and delegate.

File: solr/core/src/java/org/apache/solr/metrics/SolrCoreMetricManager.java [hunk 2]
```diff
@@ -28,6 +28,7 @@
 import org.apache.solr.core.PluginInfo;
 import org.apache.solr.core.SolrCore;
 import org.apache.solr.core.SolrInfoBean;
+import org.apache.solr.handler.RequestHandlerBase;
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/metrics/SolrCoreMetricManager.java]: unused import

File: solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler;
+
+import org.apache.solr.SolrTestCaseJ4;
+import org.apache.solr.handler.component.SearchHandler;
+import org.apache.solr.metrics.SolrDelegateRegistryMetricsContext;
+import org.apache.solr.request.SolrRequestHandler;
+import org.apache.solr.update.UpdateHandler;
```
Comment (magibney) [inline solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java]: UpdateHandler unused import

File: solr/core/src/java/org/apache/solr/metrics/SolrMetricsContext.java [hunk 4]
```diff
@@ -109,7 +109,6 @@ public Meter meter(SolrInfoBean info, String metricName, String... metricPath) {
    */
   public Counter counter(SolrInfoBean info, String metricName, String... metricPath) {
     return metricManager.counter(info, registry, metricName, metricPath);
-
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/metrics/SolrMetricsContext.java]: remove unrelated whitespace change (for clarity)

File: solr/core/src/java/org/apache/solr/metrics/DelegateRegistryHistogram.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.metrics;
+
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Reservoir;
```
Comment (magibney) [inline solr/core/src/java/org/apache/solr/metrics/DelegateRegistryHistogram.java]: unused import `Reservoir`

File: solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import org.junit.BeforeClass;
+
+/**
+ * Most of the tests for {@link org.apache.solr.handler.component.SearchHandler} are in {@link org.apache.solr.ConvertedLegacyTest}.
+ */
+public class RequestHandlerMetricsTest extends SolrTestCaseJ4 {
+  
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    initCore("solrconfig-aggreagatenodemetrics.xml", "schema.xml");
```
Comment (magibney) [inline solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java]: The test in its current state tests the configuration of the metrics context instance variables.

If we really wanted to test that the functionality works as intended, we'd probably need a SolrCloud test (to get multiple cores over which to aggregate), testing that various metrics are aggregated as expected, and that non-aggregated metrics -- e.g. gauge -- are not reflected in aggregate metrics.

I think it'd be worth making the test(s) more robust, whether or not this is considered for potential upstream contribution, because these metrics are going to be used (of course!) for making decisions, and it might not be obvious in production use if aggregation's not actually working as intended. Worst-case scenario would be drawing conclusions/making decisions based on faulty metrics! (To be clear I think this looks good now, and agree that it should work. But it would be nice to be sure, particularly across future refactoring, forward-porting to 9.x/main, etc...)

**Review by magibney — CHANGES_REQUESTED** [review]
Pending possible enhancement of tests, some minor formatting stuff, and the bug (I think?) in the delegate `Timer` metric impl, I think this looks pretty solid.

File: solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java (hunk 6 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java]: True. We should actually test the aggregate metrics are indeed returned correctly for a Solr instance with multiple cores

Comment (justinrsweeney) [reply solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java]: I updated this to test actually hitting the metrics API. At this point I can only really test counts in the unit test since QTime can be off a bit from the times captured by these metrics making it hard to validated the aggregated request times in an automated way.

**Comment by hiteshk25** [comment]
QQ: that means we need to update `PrometheusMetricsServlet` to fetch these aggregated metrics  as well?

**Comment by noblepaul** [comment]
> QQ: that means we need to update `PrometheusMetricsServlet` to fetch these aggregated metrics as well?

Is there a reason why we should not deprecate  `PrometheusMetricsServlet` and directly use the metrics API ?

**Comment by justinrsweeney** [comment]
> > QQ: that means we need to update `PrometheusMetricsServlet` to fetch these aggregated metrics as well?
> 
> Is there a reason why we should not deprecate `PrometheusMetricsServlet` and directly use the metrics API ?

In the long run I do think we want to just use the /admin/metrics endpoint as part of this epic: https://app.shortcut.com/fullstory/epic/189768?cf_workflow=500144651&ct_workflow=all&vc_group_by=day, but that'll take a bit more work to get it to prometheus. So for now I will update the current endpoint.

**Comment by justinrsweeney** [comment]
At this point, I think this is good to go, having resolved PR feedback. I have not updated the `PrometheusMetricsServlet` at this point since that isn't required to view the metrics on the nodes and that doesn't provide histogram metrics anyway.

When we migrate to just using `/admin/metrics` and the prometheus exporter, we can get the node level metrics into Prometheus and Grafana.

**Review by hiteshk25 — APPROVED** [review]
LGTM

File: solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      }
+
+      @Override
+      protected SolrResponse createResponse(SolrClient client) {
+        return new SolrResponseBase();
+      }
+    });
+
+    NamedList<Object> metrics = (NamedList<Object>) response.get("metrics");
+    NamedList<Object> nodeMetrics = (NamedList<Object>) metrics.get("solr.node");
```
Comment (magibney) [inline solr/core/src/test/org/apache/solr/handler/RequestHandlerMetricsTest.java]: Maybe check the core-level registries as well? i.e.:
```java
    metrics.forEachKey((key) -> {
      if (key.startsWith("solr.core.testRequestHandlerMetrics")) {
        NamedList<Object> nodeMetrics = (NamedList<Object>) metrics.get(key);
        assertEquals(1L, ((Map<String, Long>)nodeMetrics.get("QUERY./select.requestTimes")).get("count").longValue());
        assertEquals(1L, ((Map<String, Long>)nodeMetrics.get("UPDATE./update.requestTimes")).get("count").longValue());
      }
    });
```
As well as below (not to be overly nitpicky), would probably be better to cast to `Map<String, Long>` as opposed to `LinkedHashMap<String, Long>`, since we don't care about the specific impl type (and if the specific impl changes we could get a spurious test failure).

**Review by magibney — APPROVED** [review]
Looks good! I left one minor/optional test-related suggestion, but in any case this looks solid!
